### PR TITLE
Add Ladder Map Voting stuff

### LINKED ISSUE #52: Generate ladder map pool review vote

### PR DESCRIPTION
**Pull request by axel1200**
Fixes #52

### CODE REVIEW

**Comment by axel1200** [comment]
Attention Voting is party broken !!! on current devlop of java-api. I made the necessary change and will reference the other PR soon

**Comment by axel1200** [comment]
image

**Comment by axel1200** [comment]
https://github.com/FAForever/faf-java-api/pull/250

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 1]
```diff
@@ -93,17 +93,30 @@ public MapService(FafApiCommunicationService fafApi) {
...
+            }
+        }
+        log.trace("found {} maps", preResult.size());
         return result;
     }
 
     public void removeMapVersionFromLadderPool(MapVersion mapVersion) {
         log.debug("Deleting mapVersion from ladder pool: {}", mapVersion.getId());
-        fafApi.delete(ElideNavigator.of(Ladder1v1Map.class).id(mapVersion.getId()));
+        fafApi.delete(ElideNavigator.of(Ladder1v1Map.class).id(mapVersion.getLadder1v1Map().getId()));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: Good catch.

File: src/main/java/com/faforever/moderatorclient/api/domain/MessagesService.java [hunk 2]
```diff
@@ -33,26 +33,43 @@ public MessagesService(FafApiCommunicationService fafApi, MessagesMapper message
...
         applicationEventPublisher.publishEvent(new MessagesChangedEvent());
         return result;
     }
+
+    public Message putMessage(MessageFx messageFx) {
+        Message message = messagesMapper.map(messageFx);
+        List<Message> messages = fafApi.getAll(ElideNavigator.of(Message.class).collection()
+                .addFilter(ElideNavigator.qBuilder().string("region").eq(message.getRegion()))
+                .addFilter(ElideNavigator.qBuilder().string("language").eq(message.getLanguage()))
+                .addFilter(ElideNavigator.qBuilder().string("key").eq(message.getKey())));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MessagesService.java]: My bad. This should not be `addFilter`, but `setFilter` instead in the ElideNavigator. You can't add multiple filters, each one will override the previous.
Use the qBuilders `.and()` concatenation instead.

File: src/main/java/com/faforever/moderatorclient/api/domain/VotingService.java [hunk 3]
```diff
@@ -117,7 +120,7 @@ public VotingQuestion create(VotingQuestion votingQuestion) {
     //region choices
     public List<VotingChoice> getAllChoicesFromApi() {
         log.debug("Retrieving all choices");
-        List<VotingChoice> result = fafApi.getAll(ElideNavigator.of(VotingChoice.class).collection());
+        List<VotingChoice> result = fafApi.getAll(ElideNavigator.of(VotingChoice.class).collection().addIncludeOnCollection("votingQuestion"));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/VotingService.java]: Please follow the style as in line 79: if there are more operations put them in new lines.

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 4]
```diff
@@ -1093,7 +1092,7 @@ public static void buildMessagesTable(TableView<MessageFx> messageTableView, Mes
         messageTableView.setEditable(true);
         HashMap<TableColumn<MessageFx, ?>, Function<MessageFx, ?>> extractors = new HashMap<>();
 
-        TableColumn<MessageFx, Number> idColumn = new TableColumn<>("ID");
+        TableColumn<MessageFx, Integer> idColumn = new TableColumn<>("ID");
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: Please stick with number. There are some limitations in JavaFX where you get weird errors if you use Integer. I don't recall what is was but it is hard to track.

File: src/main/java/com/faforever/moderatorclient/ui/voting/VotingSubjectAddController.java [hunk 5]
```diff
@@ -60,21 +60,21 @@ public void onSave() {
...
+            votingSubject.setBeginOfVoteTime(beginTime);
         }
 
         OffsetDateTime endTime = OffsetDateTime.of(LocalDateTime.parse(endTimeTextField.getText(), DateTimeFormatter.ISO_LOCAL_DATE_TIME), ZoneOffset.UTC);
-        votingSubjectFX.setEndOfVoteTime(endTime);
+        votingSubject.setEndOfVoteTime(endTime);
 
         try {
-            if (votingService.create(votingSubjectFX) == null) {
+            if (votingService.create(votingSubject) == null) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/voting/VotingSubjectAddController.java]: We are separating the api entities from the fx beans on purpose. There is a strict separation between API dtos and our UI entities. Revert these changes.

File: src/main/java/com/faforever/moderatorclient/ui/main_window/VotingController.java [hunk 6]
```diff
@@ -138,12 +141,14 @@ public void addSubject() {
...
+        VotingSubject votingSubject = new VotingSubject();
         VotingSubjectFX selectedItem = subjectTable.getSelectionModel().getSelectedItem();
         if (selectedItem == null) {
             return;
         }
-        selectedItem.setRevealWinner(true);
-        votingService.updateSubject(selectedItem);
+        votingSubject.setId(selectedItem.getId());
+        votingSubject.setRevealWinner(true);
+        votingService.update(votingSubject);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/main_window/VotingController.java]: Same here. Stick with the FX beans in the controllers.

File: src/main/java/com/faforever/moderatorclient/api/domain/MessagesService.java [hunk 7]
```diff
@@ -33,26 +33,43 @@ public MessagesService(FafApiCommunicationService fafApi, MessagesMapper message
...
+        if (messages.size() > 1) {
+            throw new IllegalStateException("message should be unique by region, language and key");
+        }
+        if (!messages.isEmpty()) {
+            log.info("deleting existing message ''{}''", message);
+            messages.forEach(message1 -> deleteMessage(messagesMapper.map(message1)));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MessagesService.java]: Please us a proper variable name instead of `message1`

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 8]
```diff
@@ -1289,6 +1290,8 @@ public static void buildQuestionTable(TableView<VotingQuestionFX> tableView, Vot
         tableView.setEditable(true);
         HashMap<TableColumn<VotingQuestionFX, ?>, Function<VotingQuestionFX, ?>> extractors = new HashMap<>();
 
+        makeTableCopyable(tableView, true);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: Why are you calling it twice?

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 9]
```diff
@@ -93,17 +93,30 @@ public MapService(FafApiCommunicationService fafApi) {
...
+                })
+                .collect(Collectors.toList());
+        ArrayList<Map> result = new ArrayList<>(preResult);
+        for (Map map : preResult) {
+            //erase duplicates
+            if (result.stream().filter(map1 -> map.getId().equals(map1.getId())).count() > 1) {
+                result.remove(map);
+            }
+        }
+        log.trace("found {} maps", preResult.size());
```
[the remark this answers is not in the document]
Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: 1. The Set
As you set I use the id instead of equals because equal method won't work here, I check that. We 
might wanna change that as well.
2. Returning Map instead of MapVersion
I changed it because of problems that were there. You add one map version, the map was added with all its versions and version that were not included were shown empty. And if you added two versions of the same map the map was shown twice. So I corrected all those fallacies. I only show versions that are in the ladder map pool and I don't show map duplicates any longer. From the way the tree view is structured it needs Map objects and also I did not change the return type I only corrected the data output of the method.

[the remark this answers is not in the document]
Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: I don't think I should change anything here should I?

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java (hunk 1 above)
Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: thx

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java (hunk 8 above)
Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: I am not calling it twice! Please unfold the code completely. When looking at the code spinet it seems. but the two calls are in different methods...

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Brutus5000) [reply src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: Oh that was due to some Lombok annotation leftovers. Fixed in https://github.com/FAForever/faf-java-commons/commit/6623f1f59c8bcb0d7c33423708cbf169e7c65638

File: src/main/java/com/faforever/moderatorclient/api/domain/MapService.java [hunk 10]
```diff
@@ -93,17 +94,23 @@ public MapService(FafApiCommunicationService fafApi) {
...
 
-        List<Map> result = ladder1v1Maps.stream()
+        List<MapVersion> ladder1v1MapVersions = ladder1v1Maps.stream()
                 .map(Ladder1v1Map::getMapVersion)
-                .map(MapVersion::getMap)
                 .collect(Collectors.toList());
+        Set<Map> result = ladder1v1MapVersions.stream()
+                .map(mapVersion -> {
+                    Map map = mapVersion.getMap();
+                    map.setVersions(map.getVersions().stream().filter(ladder1v1MapVersions::contains).collect(Collectors.toList()));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/domain/MapService.java]: Proper formatting now an we're good.

**Comment by axel1200** [comment]
like this

**Comment by Brutus5000** [comment]
Should this work against develop-api? Creating a vote fails with an error without using the regular exception message bo and then it throws a success message afterwards. Both needs to be fixed.

**Comment by axel1200** [comment]
As I wrote early no it does not work against develop api

**Comment by axel1200** [comment]
FAForever/faf-java-api#250 depends on

**Comment by axel1200** [comment]
fixed that false positive dialog

File: src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java [hunk 11]
```diff
@@ -84,6 +83,8 @@ public FafApiCommunicationService(ResourceConverter resourceConverter, Applicati
                 .additionalMessageConverters(jsonApiMessageConverter)
                 .errorHandler(jsonApiErrorHandler)
                 .rootUri(apiBaseUrl);
+        headers = new HttpHeaders();
```
Comment (axel1200) [inline src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: @bukajsytlos @Brutus5000  for post i got headers working, however i still need to due it for patch. Do you know if there is a way to set headers in general?

File: src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java [hunk 12]
```diff
@@ -149,9 +151,13 @@ public LegacyAccessLevel login(String username, String password) {
...
+        httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+        httpHeaders.setContentType(MediaType.APPLICATION_JSON);
+
         try {
             authorizedLatch.await();
-            return restOperations.exchange(url, HttpMethod.PATCH, new HttpEntity<>(object), routeBuilder.getDtoClass()).getBody();
+            return restOperations.exchange(url, HttpMethod.PATCH, new HttpEntity<>(object, httpHeaders), routeBuilder.getDtoClass()).getBody();
         } catch (Throwable t) {
             applicationEventPublisher.publishEvent(new FafApiFailModifyEvent(t, routeBuilder.getDtoClass(), url));
             throw t;
```
Comment (axel1200) [inline src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: How do i do it for the patch method blow @bukajsytlos @Brutus5000

File: src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java [hunk 13]
```diff
@@ -99,7 +104,14 @@ private void authorize(String username, String password) {
         details.setUsername(username);
         details.setPassword(password);
 
-        restOperations = restTemplateBuilder.configure(new OAuth2RestTemplate(details));
+        restTemplate = restTemplateBuilder.configure(new OAuth2RestTemplate(details));
+        restTemplate.setInterceptors(Collections.singletonList(
```
Comment (bukajsytlos) [inline src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: I guess this will set it globally overriding anything we set in request right? Do we really want this? I would prefer to just modify  POST and PATCH methods.

Comment (Brutus5000) [reply src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: The only place where we do not send application/json is the login (form encoded). Even there the code works. I see no reason to set it per method.

Comment (bukajsytlos) [reply src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: I mean something like this
```
        restTemplate.setInterceptors(Collections.singletonList(
                (request, body, execution) -> {
                    if (request.getMethod() == HttpMethod.POST || request.getMethod() == HttpMethod.PATCH || request.getMethod() == HttpMethod.PUT) {
                        HttpHeaders headers = request.getHeaders();
                        headers.setContentType(MediaType.APPLICATION_JSON);
                    }
                    return execution.execute(request, body);
                }
        ));
```

Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: I would agree to @bukajsytlos just because it works does not mean we should do it. So if in the future somebody changes the MeidaTypes in the api he can be sure they are all set correctly.

**Comment by axel1200** [comment]
The changes you made do not work @Brutus5000 

```
2018-09-22 20:27:07.396  WARN 520912 --- [tp2074687317-93] c.f.a.e.GlobalControllerExceptionHandler : Internal server error

org.springframework.web.HttpMediaTypeNotAcceptableException: Could not find acceptable representation
	at org.springframework.web.servlet.mvc.method.RequestMappingInfoHandlerMapping.handleNoMatch(RequestMappingInfoHandlerMapping.java:216)
	at org.springframework.web.servlet.handler.AbstractHandlerMethodMapping.lookupHandlerMethod(AbstractHandlerMethodMapping.java:376)
	at org.springframework.web.servlet.handler.AbstractHandlerMethodMapping.getHandlerInternal(AbstractHandlerMethodMapping.java:316)
	at org.springframework.web.servlet.handler.AbstractHandlerMethodMapping.getHandlerInternal(AbstractHandlerMethodMapping.java:62)
	at org.springframework.web.servlet.handler.AbstractHandlerMapping.getHandler(AbstractHandlerMapping.java:350)
	at org.springframework.web.servlet.DispatcherServlet.getHandler(DispatcherServlet.java:1188)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:964)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:925)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:974)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:877)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:851)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:215)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.oauth2.provider.authentication.OAuth2AuthenticationProcessingFilter.doFilter(OAuth2AuthenticationProcessingFilter.java:176)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by axel1200** [comment]
I would supect it is because you only set the content type but not what it consumes 🤔

**Comment by axel1200** [comment]
```
 @throws HttpMediaTypeNotAcceptableException if there are matches by URL
	 * but not by consumable/producible media types
```

**Comment by axel1200** [comment]
image
image
Everything matches I dont get it. @bukajsytlos got an idea?

**Comment by axel1200** [comment]
Works when I do this 
image

**Comment by Brutus5000** [comment]
When do you run into the error? Is the stacktrace API or client?

**Comment by axel1200** [comment]
API stacktrace

**Comment by axel1200** [comment]
I think your changes are good @bukajsytlos I will test it

**Comment by axel1200** [comment]
Works fine however ladder vote broke but i am sure i can quickly fix that

**Comment by axel1200** [comment]
All good to go and finally working again

File: src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java (hunk 13 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/moderatorclient/api/FafApiCommunicationService.java]: The formatting is ugly, but I accept this as the better solution.
